src/sentry/migrations/0552_add_team_replica.py

```diff
@@ -0,0 +1,61 @@
+# Generated by Django 3.2.20 on 2023-09-14 03:21
+
+import django.utils.timezone
+from django.db import migrations, models
+
+import sentry.db.models.fields.bounded
+import sentry.db.models.fields.hybrid_cloud_foreign_key
+from sentry.new_migrations.migrations import CheckedMigration
+
+
+class Migration(CheckedMigration):
+    # This flag is used to mark that a migration shouldn't be automatically run in production. For
+    # the most part, this should only be used for operations where it's safe to run the migration
+    # after your code has deployed. So this should not be used for most operations that alter the
+    # schema of a table.
+    # Here are some things that make sense to mark as dangerous:
+    # - Large data migrations. Typically we want these to be run manually by ops so that they can
+    #   be monitored and not block the deploy for a long period of time while they run.
+    # - Adding indexes to large tables. Since this can take a long time, we'd generally prefer to
+    #   have ops run this and not block the deploy. Note that while adding an index is a schema
+    #   change, it's completely safe to run the operation after the code has deployed.
+    is_dangerous = False
+
+    dependencies = [
+        ("sentry", "0551_drop_xactor_actor"),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name="TeamReplica",
+            fields=[
+                (
+                    "id",
+                    sentry.db.models.fields.bounded.BoundedBigAutoField(
+                        primary_key=True, serialize=False
+                    ),
+                ),
+                (
+                    "team_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.Team", db_index=True, on_delete="CASCADE"
+                    ),
+                ),
+                (
+                    "organization_id",
+                    sentry.db.models.fields.hybrid_cloud_foreign_key.HybridCloudForeignKey(
+                        "sentry.Organization", db_index=True, on_delete="CASCADE"
+                    ),
+                ),
+                ("slug", models.SlugField()),
+                ("name", models.CharField(max_length=64)),
+                ("status", sentry.db.models.fields.bounded.BoundedPositiveIntegerField()),
+                ("date_added", models.DateTimeField(default=django.utils.timezone.now)),
+                ("org_role", models.CharField(max_length=32, null=True)),
+            ],
+            options={
+                "db_table": "sentry_teamreplica",
+                "unique_together": {("organization_id", "slug")},
```

Comment:
Couldn't this make handling renames tricky? For example if a user swaps the names of two teams in the UI 
(`alpha -> charlie`, `beta -> alpha` and `charlie -> beta`),  when processing the outboxes for these three operations we could process the beta->alpha change before the alpha->charlie one.

Without the unique index we are free to replicate the data in the order it is given to control and let eventual consistency take the wheel.

Reply:
I could see us having a unique_constraint on `organization_id, team_id` as duplicate records are an invalid state that needs to be prevented.

Reply:
> Great question, and I have been thinking deeply about this problem, and I have a general solution

:heart:

> The reason we DO want the unique columns is that it supports the query assumptions that back those columns. Someone reading TeamReplica and getting multiple teams with the same slug for the same org has to do extra logic to interpret what it means.

I hadn't considered the read side semantics being changed as queries that used to find 0..1 records could now get 0..N records.

> But someone reading TeamReplica that experiences short team "blip" in which one element disappears and reappears shortly doesn't need to perform any extra logical work

I was thinking of the blip as the larger problem vs having to handle the conflicts before they happen. However, the impact of changing the read side semantics are more important than avoiding a missing team blip when users contrive conflicts like this.